tests/components/mqtt/test_init.py

```diff
@@ -1135,46 +1135,71 @@ async def test_logs_error_if_no_connect_broker(
     )
 
 
-async def test_handle_mqtt_on_callback(hass, caplog, mqtt_mock, mqtt_client_mock):
-    """Test handling of an ACK callback."""
-    mqtt_mock()._pending_operations = {}
-
-    mid = 1
-    mqtt_client_mock.on_publish(mqtt_client_mock, None, mid)
-    # mqtt_mock._mqtt_handle_mid(mid)
+@patch("homeassistant.components.mqtt.TIMEOUT_ACK", 0.3)
+async def test_publish_mid(hass, mqtt_client_mock):
+    """Test if subscribe, publish, and unsubscribe waits until the callback release."""
+    unsub_1 = None
+    unsub_2 = None
+    calls_1 = MagicMock()
+    calls_2 = MagicMock()
+
+    # on_publish: no callback yet
+    async def publish_twice():
+        """Publish two calls without a simulated callback."""
+        await mqtt.async_publish(hass, "test-topic1", "payload1")
+        await mqtt.async_publish(hass, "no_callback/test-topic2", "payload2")
+
+    # on_subscribe: no callback yet
+    async def subscribe_twice():
+        """Publish two calls without a simulated callback."""
+        nonlocal unsub_1, unsub_2, calls_1, calls_2
+        unsub_1 = await mqtt.async_subscribe(hass, "test-topic1", calls_1)
+        unsub_2 = await mqtt.async_subscribe(hass, "no_callback/test-topic2", calls_2)
+
+    # on_unsubscribe: no callback yet
+    async def unsubscribe_twice():
+        """Publish two calls without a simulated callback."""
+        nonlocal unsub_1, unsub_2
+        unsub_1()
+        unsub_2()
+
+    # Setup
+    entry = MockConfigEntry(domain=mqtt.DOMAIN, data={mqtt.CONF_BROKER: "test-broker"})
+    assert await mqtt.async_setup_entry(hass, entry)
     await hass.async_block_till_done()
-    await asyncio.sleep(0.2)
-    assert mid in mqtt_mock()._pending_operations
-    del mqtt_mock()._pending_operations[mid]
+    # We are connected
+    mqtt_client_mock.on_connect(mqtt_client_mock, None, None, 0)
 
-    mid = 2
-    mqtt_client_mock.on_subscribe(mqtt_client_mock, None, mid)
-    # mqtt_mock._mqtt_handle_mid(mid)
+    # Subscribe on two topics the first with ACK
+    hass.async_create_task(subscribe_twice())
     await hass.async_block_till_done()
-    await asyncio.sleep(0.2)
-    assert mid in mqtt_mock()._pending_operations
-    del mqtt_mock()._pending_operations[mid]
-
-    mid = 3
-    mqtt_client_mock.on_unsubscribe(mqtt_client_mock, None, mid)
-    # mqtt_mock._mqtt_handle_mid(mid)
+    # Publish on two topics the first with ACK
+    hass.async_create_task(publish_twice())
+    # Subscribe on two topics the first with ACK
+    hass.async_create_task(unsubscribe_twice())
     await hass.async_block_till_done()
-    await asyncio.sleep(0.2)
-    assert mid in mqtt_mock()._pending_operations
-    del mqtt_mock()._pending_operations[mid]
-
+    # Only the first published call was acknowlegded
+    assert calls_1.call_count == 1
+    assert calls_2.call_count == 0
```

Comment:
This test isn't completely testing the publish logic that says that the publish call should wait for the publish operation completion. The test is also quite complex since we're involving the subscription logic. I recommend using my example approach instead where we control the operation callback explicitly and can assert the call progress at each step.

Reply:
I think it does, though I had in mind the improve the assertion a bit though. Besides publishing also subscribing and unsubscribing expects an ACK callback from the client. At the calling line _wait_for_mid (mid is returned from the client). The client will fire a callback resulting in a call to _mqtt_handle_mid. If _mqtt_handle_mid is called before waiting, the actial waiting is not tested. What I want to see is that `_wait_for_mid is called without `_mqtt_handle_mid` being fired yet. That should cause a timeout. To set the event later, I need the mid value, which is not exposed. `test_handle_mqtt_on_callback` tests the case where the event was set before wait was called.

Reply:
I think `test_wait_for_ack` (just added) actually tests that.